blackjax/vi/fullrank_vi.py

```diff
@@ -0,0 +1,188 @@
+
+
+def as_top_level_api(
+    logdensity_fn: Callable,
+    optimizer: GradientTransformation,
+    num_samples: int = 100,
+):
+    """High-level implementation of Full-Rank Variational Inference.
+
+    Parameters
+    ----------
+    logdensity_fn
+        A function that represents the log-density function associated with
+        the distribution we want to sample from.
+    optimizer
+        Optax optimizer to use to optimize the ELBO.
+    num_samples
+        Number of samples to take at each step to optimize the ELBO.
+
+    Returns
+    -------
+    A ``VIAlgorithm``.
+
+    """
+
+    def init_fn(position: ArrayLikeTree):
+        return init(position, optimizer)
+
+    def step_fn(rng_key: PRNGKey, state: FRVIState) -> tuple[FRVIState, FRVIInfo]:
+        return step(rng_key, state, logdensity_fn, optimizer, num_samples)
+
+    def sample_fn(rng_key: PRNGKey, state: FRVIState, num_samples: int):
+        return sample(rng_key, state, num_samples)
+
+    return VIAlgorithm(init_fn, step_fn, sample_fn)
+
+
+def _unflatten_cholesky(chol_params):
+    """Construct the Cholesky factor from a flattened vector of cholesky parameters.
+
+    Transforms a flattened vector representation of a lower triangular matrix
+    into a full Cholesky factor. The input vector contains n = d(d+1)/2 elements
+    consisting of d diagonal elements followed by n - d off-diagonal elements in
+    row-major order, where d is the dimension of the matrix.
+
+    The diagonal elements are passed through a softplus function to ensure (numerically
+    stable) positivity, crucial to maintain a valid covariance matrix parameterization.
+
+    This parameterization allows for unconstrained optimization while ensuring the
+    resulting covariance matrix Sigma = CC^T is symmetric and positive definite.
+    """
+    n = chol_params.size
+    dim = int(jnp.sqrt(1 + 8 * n) - 1) // 2
+    tril = jnp.zeros((dim, dim))
+    tril = tril.at[jnp.tril_indices(dim, k=-1)].set(chol_params[dim:])
+    diag = jax.nn.softplus(chol_params[:dim])
+    chol_factor = tril + jnp.diag(diag)
+    return chol_factor
+
+
+def _sample(rng_key, mu, chol_params, num_samples):
+    mu_flatten, unravel_fn = jax.flatten_util.ravel_pytree(mu)
+    chol_factor = _unflatten_cholesky(chol_params)
+    eps = jax.random.normal(rng_key, (num_samples, mu_flatten.shape[0]))
+    flatten_sample = eps @ chol_factor.T + mu_flatten
+    return jax.vmap(unravel_fn)(flatten_sample)
+
+
+def generate_fullrank_logdensity(mu, chol_params):
+    mu_flatten, _ = jax.flatten_util.ravel_pytree(mu)
+    chol_factor = _unflatten_cholesky(chol_params)
+    cov = chol_factor @ chol_factor.T
+
+    def fullrank_logdensity(position):
+        position_flatten = jax.flatten_util.ravel_pytree(position)[0]
+        # TODO: inefficient because of redundant cholesky decomposition
+        return jsp.stats.multivariate_normal.logpdf(position_flatten, mu_flatten, cov)
```

Comment:
You have a good point about computation efficiency, let's keep the cholesky version of the logpdf. Could you rewrite the API to similar to `multivariate_normal.logpdf`, and use function partial when you call it?

Reply:
Can you specify the API you imagine my implementation would have and how function partial would be applied to it? Feel free to reference my code I just committed.